tracer/src/Datadog.Trace/Configuration/ConfigurationSources/Telemetry/ConfigurationBuilder.cs

```diff
@@ -0,0 +1,300 @@
+﻿// <copyright file="ConfigurationBuilder.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Datadog.Trace.Configuration.ConfigurationSources.Telemetry;
+
+namespace Datadog.Trace.Configuration.Telemetry;
+
+internal readonly struct ConfigurationBuilder
+{
+    private readonly ITelemeteredConfigurationSource _source;
+    private readonly IConfigurationTelemetry _telemetry;
+
+    public ConfigurationBuilder(IConfigurationSource source, IConfigurationTelemetry telemetry)
+    {
+        // If the source _isn't_ an ITelemeteredConfigurationSource, it's because it's a custom
+        // IConfigurationSource implementation, so we treat that as a "Code" origin.
+        _source = source as ITelemeteredConfigurationSource ?? new CustomTelemeteredConfigurationSource(source);
+        _telemetry = telemetry;
+    }
+
+    public HasKeys WithKeys(string key) => new(_source, _telemetry, key);
+
+    public HasKeys WithKeys(string key, string fallbackKey) => new(_source, _telemetry, key, fallbackKey);
+
+    public HasKeys WithKeys(string key, string fallbackKey1, string fallbackKey2) => new(_source, _telemetry, key, fallbackKey1, fallbackKey2);
+
+    public HasKeys WithKeys(string key, string fallbackKey1, string fallbackKey2, string fallbackKey3) => new(_source, _telemetry, key, fallbackKey1, fallbackKey2, fallbackKey3);
+
+    internal readonly struct HasKeys
+    {
+        public HasKeys(ITelemeteredConfigurationSource source, IConfigurationTelemetry telemetry, string key, string? fallbackKey1 = null, string? fallbackKey2 = null, string? fallbackKey3 = null)
+        {
+            Source = source;
+            Telemetry = telemetry;
+            Key = key;
+            FallbackKey1 = fallbackKey1;
+            FallbackKey2 = fallbackKey2;
+            FallbackKey3 = fallbackKey3;
+        }
+
+        private ITelemeteredConfigurationSource Source { get; }
+
+        private IConfigurationTelemetry Telemetry { get; }
+
+        private string Key { get; }
+
+        private string? FallbackKey1 { get; }
+
+        private string? FallbackKey2 { get; }
+
+        private string? FallbackKey3 { get; }
+
+        public StringAccessor AsString() => new(this, recordValue: true);
```

Comment:
I don't understand why this extra indirection (`StringAccessor`, `BoolAccessor`, ...) instead of directly `GetAsString`, `GetAsBool`, ...

Reply:
Yeah, looks like this is a symptom of many refactors. Originally I was trying to lean into the builder pattern hard, something like:

```csharp
.AsString()
.WithDefaultValue(value)
.WithValidation()
.Get()`
```

but that ran into a bunch of other issues, hence where we are. Pretty sure I can remove that indirection

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/4033/commits/11cd46f3c79dc5fd7d473bace3edb3c5c4e7ef37